File: scripts/comprehensive_analysis.py
#!/usr/bin/env python3
"""
Umfassende Analyse: Species Richness + MaxN + Vergleichskennzahlen
Kombiniert alle Annotation Reports (Coral Reef + Nursery)
Erstellt große Vergleichstabelle mit Species Richness, MaxN und statistischen Kennzahlen
"""

import pandas as pd
import numpy as np
from pathlib import Path
from scipy.stats import entropy
from collections import Counter

# ===== 1. FUNKTIONEN DEFINIEREN =====

def calculate_shannon_index(species_list):
    """Berechnet Shannon Diversity Index"""
    if len(species_list) == 0:
        return 0
    counts = Counter(species_list)
    total = sum(counts.values())
    proportions = [count / total for count in counts.values()]
    return entropy(proportions)

def calculate_simpson_index(species_list):
    """Berechnet Simpson Diversity Index"""
    if len(species_list) == 0:
        return 0
    counts = Counter(species_list)
    total = sum(counts.values())
    proportions = [count / total for count in counts.values()]
    return 1 - sum(p**2 for p in proportions)

def calculate_pielou_evenness(species_list):
    """Berechnet Pielou's Evenness Index (J)"""
    unique_species = len(set(species_list))
    if unique_species <= 1:
        return 0
    shannon = calculate_shannon_index(species_list)
    return shannon / np.log(unique_species)

def calculate_comprehensive_stats(file_path, file_name, area_type):
    """Berechnet comprehensive Statistics für eine Datei"""
    try:
        # Daten einlesen
        data = pd.read_csv(file_path)
        
        if len(data) == 0:
            print(f"Warnung: Datei leer - {file_name}")
            return None
        
        # Basic Statistics
        unique_species = data['label_name'].nunique()
        total_observations = len(data)
        
        # MaxN Berechnung: Höchste Anzahl einer Art in einem Frame
        maxn_by_species = data.groupby(['label_name', 'frames']).size().reset_index(name='count')
        overall_maxn = maxn_by_species['count'].max()
        
        # Beobachtungen pro Art
        observations_per_species = data.groupby('label_name').size()
        obs_per_species_mean = observations_per_species.mean()
        obs_per_species_sd = observations_per_species.std()
        
        # Diversity Indices
        species_list = data['label_name'].tolist()
        shannon_h = calculate_shannon_index(species_list)
        simpson_d = calculate_simpson_index(species_list)
        pielou_j = calculate_pielou_evenness(species_list)
        
        # Anzahl der Frames im Video
        unique_frames = data['frames'].nunique()
        
        # Top 3 häufigste Arten
        top_species = observations_per_species.nlargest(3).index.tolist()
        top_species_str = ", ".join(top_species[:3])
        
        # Informationen aus Dateinamen extrahieren
        location = "Utumbi" if "utumbi" in file_name.lower() else "Milimani"
        date_str = file_name.split('-')[0]
        
        # Datum formatieren (DDMMJJJJ -> DD.MM.JJJJ)
        if len(date_str) == 8 and date_str.isdigit():
            date_formatted = f"{date_str[0:2]}.{date_str[2:4]}.{date_str[4:8]}"
        else:
            date_formatted = date_str
        
        # Köder/Bait Information
        import re
        match = re.search(r'-c\d+-(.+?)-ganz\.csv$', file_name.lower())
        if match:
            bait_info = match.group(1)
        else:
            match = re.search(r'-(.+?)-ganz\.csv$', file_name.lower())
            bait_info = match.group(1) if match else "unknown"
        
        # Ergebnis als Dictionary
        result = {
            'file_name': file_name,
            'area_type': area_type,
            'location': location,
            'date': date_formatted,
            'bait_type': bait_info,
            'species_richness': unique_species,
            'total_observations': total_observations,
            'maxn_overall': overall_maxn,
            'obs_per_species_mean': round(obs_per_species_mean, 2),
            'obs_per_species_sd': round(obs_per_species_sd, 2),
            'shannon_h': round(shannon_h, 3),
            'simpson_d': round(simpson_d, 3),
            'pielou_j': round(pielou_j, 3),
            'unique_frames': unique_frames,
            'obs_per_frame': round(total_observations / unique_frames, 2),
            'top_3_species': top_species_str
        }
        
        return result
        
    except Exception as e:
        print(f"Fehler bei Datei: {file_name} - {str(e)}")
        return None

# ===== 2. ALLE DATEIEN SAMMELN UND ANALYSIEREN =====

print("\n===== ANALYSIERE ALLE ANNOTATION REPORTS =====\n")

coral_reef_dir = Path("Annotation_reports_coral_reef")
nursery_dir = Path("Annotation_reports_Nursery")

# Finde alle CSV-Dateien
coral_reef_files = list(coral_reef_dir.glob("*.csv"))
nursery_files = list(nursery_dir.glob("*.csv"))

print(f"Coral Reef Dateien gefunden: {len(coral_reef_files)}")
print(f"Nursery Dateien gefunden: {len(nursery_files)}\n")

all_results = []

# Analysiere Coral Reef
print("Analysiere Coral Reef Dateien...")
for file in sorted(coral_reef_files):
    file_name = file.name
    print(f"  - {file_name}")
    result = calculate_comprehensive_stats(str(file), file_name, "Coral Reef")
    if result:
        all_results.append(result)

# Analysiere Nursery
print("\nAnalysiere Nursery Dateien...")
for file in sorted(nursery_files):
    file_name = file.name
    print(f"  - {file_name}")
    result = calculate_comprehensive_stats(str(file), file_name, "Nursery")
    if result:
        all_results.append(result)

# ===== 3. ERSTELLE HAUPTTABELLE =====

comprehensive_df = pd.DataFrame(all_results)

# Sortiere nach Species Richness (absteigend)
comprehensive_df = comprehensive_df.sort_values('species_richness', ascending=False).reset_index(drop=True)

# ===== 4. ERGEBNISSE ANZEIGEN =====

print("\n\n===== UMFASSENDE VERGLEICHSTABELLE =====\n")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)
print(comprehensive_df.to_string())

# ===== 5. STATISTIKEN NACH BEREICH =====

print("\n\n===== STATISTIKEN NACH BEREICH =====\n")

stats_by_area = comprehensive_df.groupby('area_type').agg({
    'file_name': 'count',
    'species_richness': ['mean', 'std', 'min', 'max'],
    'maxn_overall': ['mean', 'std'],
    'total_observations': 'sum',
    'shannon_h': 'mean',
    'pielou_j': 'mean'
}).round(2)

stats_by_area.columns = ['anzahl_dateien', 'richness_mean', 'richness_sd', 'richness_min', 
                         'richness_max', 'maxn_mean', 'maxn_sd', 'observations_total', 
                         'shannon_mean', 'evennes_mean']
print(stats_by_area)

# ===== 6. STATISTIKEN NACH STANDORT =====

print("\n\n===== STATISTIKEN NACH STANDORT =====\n")

stats_by_location = comprehensive_df.groupby(['location', 'area_type']).agg({
    'file_name': 'count',
    'species_richness': ['mean', 'std'],
    'maxn_overall': 'mean',
    'total_observations': 'sum',
    'shannon_h': 'mean'
}).round(2)

stats_by_location.columns = ['anzahl_dateien', 'richness_mean', 'richness_sd', 
                              'maxn_mean', 'observations_total', 'shannon_mean']
print(stats_by_location)

# ===== 7. TOP KÖDERTYPEN =====

print("\n\n===== TOP KÖDERTYPEN (Bait Types) =====\n")

bait_stats = comprehensive_df.groupby('bait_type').agg({
    'file_name': 'count',
    'species_richness': ['mean', 'std'],
    'maxn_overall': 'mean',
    'total_observations': 'mean',
    'shannon_h': 'mean'
}).round(2)

bait_stats.columns = ['anzahl_videos', 'richness_mean', 'richness_sd', 
                      'maxn_mean', 'obs_mean', 'shannon_mean']
bait_stats = bait_stats.sort_values('richness_mean', ascending=False)
print(bait_stats)

# ===== 8. SPEICHERE ERGEBNISSE =====

output_dir = Path("results")
output_dir.mkdir(exist_ok=True)

# Haupttabelle
comprehensive_df.to_csv(output_dir / "comprehensive_analysis_table.csv", index=False)
print("\n\nHaupttabelle gespeichert: results/comprehensive_analysis_table.csv")

# Bereichs-Statistiken
stats_by_area.to_csv(output_dir / "statistics_by_area.csv")
print("Bereichs-Statistiken gespeichert: results/statistics_by_area.csv")

# Standort-Statistiken
stats_by_location.to_csv(output_dir / "statistics_by_location.csv")
print("Standort-Statistiken gespeichert: results/statistics_by_location.csv")

# Köder-Statistiken
bait_stats.to_csv(output_dir / "statistics_by_bait.csv")
print("Köder-Statistiken gespeichert: results/statistics_by_bait.csv")

print("\n===== ANALYSE ABGESCHLOSSEN =====\n")

File: scripts/comprehensive_analysis.R
# ===== Umfassende Analyse: Species Richness + MaxN + Vergleichskennzahlen =====
# Kombiniert alle Annotation Reports (Coral Reef + Nursery)
# Erstellt große Vergleichstabelle mit Species Richness, MaxN und statistischen Kennzahlen

library(dplyr)
library(tidyr)
library(vegan)

# ===== 1. FUNKTIONEN DEFINIEREN =====

# Berechnet comprehensive Statistics für eine Datei
calculate_comprehensive_stats <- function(file_path, file_name, area_type) {
  tryCatch({
    # Daten einlesen
    data <- read.csv(file_path, stringsAsFactors = FALSE)
    
    if (nrow(data) == 0) {
      cat("Warnung: Datei leer -", file_name, "\n")
      return(NULL)
    }
    
    # Basic Statistics
    unique_species <- n_distinct(data$label_name)
    total_observations <- nrow(data)
    
    # MaxN Berechnung für jede Art
    maxn_by_species <- data %>%
      group_by(label_name, frames) %>%
      summarise(count = n(), .groups = "drop") %>%
      group_by(label_name) %>%
      summarise(MaxN = max(count), .groups = "drop")
    
    # Gesamtes MaxN (höchste Anzahl einer Art in einem Frame)
    overall_maxn <- max(maxn_by_species$MaxN)
    
    # Beobachtungen pro Art
    observations_per_species <- data %>%
      group_by(label_name) %>%
      summarise(count = n(), .groups = "drop")
    
    obs_per_species_mean <- mean(observations_per_species$count)
    obs_per_species_sd <- sd(observations_per_species$count)
    
    # Shannon Diversity Index
    species_freq <- table(data$label_name)
    shannon_h <- diversity(species_freq, index = "shannon")
    
    # Simpson Diversity Index
    simpson_d <- diversity(species_freq, index = "simpson")
    
    # Evenness (Pielou's J)
    pielou_j <- shannon_h / log(unique_species)
    
    # Anzahl der Frames im Video
    unique_frames <- n_distinct(data$frames)
    
    # Top 3 häufigste Arten
    top_species <- observations_per_species %>%
      arrange(desc(count)) %>%
      head(3) %>%
      pull(label_name) %>%
      paste(collapse = ", ")
    
    # Informationen aus Dateinamen extrahieren
    location <- ifelse(grepl("utumbi", tolower(file_name)), "Utumbi", "Milimani")
    date <- sub("^(\\d+)-.*", "\\1", file_name)
    
    # Datum formatieren (DDMMJJJJ -> DD.MM.JJJJ)
    if (nchar(date) == 8) {
      date_formatted <- paste0(substr(date, 1, 2), ".", substr(date, 3, 4), ".", substr(date, 5, 8))
    } else {
      date_formatted <- date
    }
    
    # Köder/Bait Information
    bait_info <- sub(".*-c\\d+-(.+)-ganz\\.csv$", "\\1", tolower(file_name))
    if (!grepl("-", file_name)) {
      bait_info <- sub(".*-(.+)-ganz\\.csv$", "\\1", tolower(file_name))
    }
    
    # Ergebnis als Dataframe
    result <- data.frame(
      file_name = file_name,
      area_type = area_type,
      location = location,
      date = date_formatted,
      bait_type = bait_info,
      species_richness = unique_species,
      total_observations = total_observations,
      maxn_overall = overall_maxn,
      obs_per_species_mean = round(obs_per_species_mean, 2),
      obs_per_species_sd = round(obs_per_species_sd, 2),
      shannon_h = round(shannon_h, 3),
      simpson_d = round(simpson_d, 3),
      pielou_j = round(pielou_j, 3),
      unique_frames = unique_frames,
      obs_per_frame = round(total_observations / unique_frames, 2),
      top_3_species = top_species,
      stringsAsFactors = FALSE
    )
    
    return(result)
    
  }, error = function(e) {
    cat("Fehler bei Datei:", file_name, "-", e$message, "\n")
    return(NULL)
  })
}

# ===== 2. ALLE DATEIEN SAMMELN =====

cat("\n===== ANALYSIERE ALLE ANNOTATION REPORTS =====\n\n")

# Coral Reef Dateien
coral_reef_dir <- "Annotation_reports_coral_reef"
coral_reef_files <- list.files(coral_reef_dir, pattern = "\\.csv$", full.names = TRUE)

# Nursery Dateien
nursery_dir <- "Annotation_reports_Nursery"
nursery_files <- list.files(nursery_dir, pattern = "\\.csv$", full.names = TRUE)

cat("Coral Reef Dateien gefunden:", length(coral_reef_files), "\n")
cat("Nursery Dateien gefunden:", length(nursery_files), "\n\n")

# ===== 3. ANALYSIERE ALLE DATEIEN =====

all_results <- list()

# Coral Reef
cat("Analysiere Coral Reef Dateien...\n")
for (file in coral_reef_files) {
  file_name <- basename(file)
  cat("  -", file_name, "\n")
  result <- calculate_comprehensive_stats(file, file_name, "Coral Reef")
  if (!is.null(result)) {
    all_results[[length(all_results) + 1]] <- result
  }
}

# Nursery
cat("\nAnalysiere Nursery Dateien...\n")
for (file in nursery_files) {
  file_name <- basename(file)
  cat("  -", file_name, "\n")
  result <- calculate_comprehensive_stats(file, file_name, "Nursery")
  if (!is.null(result)) {
    all_results[[length(all_results) + 1]] <- result
  }
}

# ===== 4. ZUSAMMENFASSUNG IN DATAFRAME =====

comprehensive_df <- do.call(rbind, all_results)
rownames(comprehensive_df) <- NULL

# Sortieren nach Species Richness (absteigend)
comprehensive_df <- comprehensive_df %>%
  arrange(desc(species_richness))

# ===== 5. ERGEBNISSE ANZEIGEN =====

cat("\n\n===== UMFASSENDE VERGLEICHSTABELLE =====\n\n")
print(comprehensive_df, n = Inf)

# ===== 6. STATISTIKEN NACH BEREICH =====

cat("\n\n===== STATISTIKEN NACH BEREICH =====\n\n")

stats_by_area <- comprehensive_df %>%
  group_by(area_type) %>%
  summarise(
    anzahl_dateien = n(),
    richness_mean = round(mean(species_richness), 2),
    richness_sd = round(sd(species_richness), 2),
    richness_min = min(species_richness),
    richness_max = max(species_richness),
    maxn_mean = round(mean(maxn_overall), 2),
    maxn_sd = round(sd(maxn_overall), 2),
    observations_total = sum(total_observations),
    shannon_mean = round(mean(shannon_h), 3),
    evennes_mean = round(mean(pielou_j), 3),
    .groups = "drop"
  )

print(stats_by_area)

# ===== 7. STATISTIKEN NACH STANDORT (Location) =====

cat("\n\n===== STATISTIKEN NACH STANDORT =====\n\n")

stats_by_location <- comprehensive_df %>%
  group_by(location, area_type) %>%
  summarise(
    anzahl_dateien = n(),
    richness_mean = round(mean(species_richness), 2),
    richness_sd = round(sd(species_richness), 2),
    maxn_mean = round(mean(maxn_overall), 2),
    observations_total = sum(total_observations),
    shannon_mean = round(mean(shannon_h), 3),
    .groups = "drop"
  )

print(stats_by_location)

# ===== 8. TOP KÖDERTYPEN =====

cat("\n\n===== TOP KÖDERTYPEN (Bait Types) =====\n\n")

bait_stats <- comprehensive_df %>%
  group_by(bait_type) %>%
  summarise(
    anzahl_videos = n(),
    richness_mean = round(mean(species_richness), 2),
    richness_sd = round(sd(species_richness), 2),
    maxn_mean = round(mean(maxn_overall), 2),
    obs_mean = round(mean(total_observations), 0),
    shannon_mean = round(mean(shannon_h), 3),
    .groups = "drop"
  ) %>%
  arrange(desc(richness_mean))

print(bait_stats)

# ===== 9. STATISTISCHE SIGNIFIKANZTESTS =====

cat("\n\n===== STATISTISCHE SIGNIFIKANZTESTS =====\n\n")

significance_results <- data.frame()

# Test 1: Coral Reef vs Nursery (Species Richness)
coral_reef_richness <- comprehensive_df %>%
  filter(area_type == "Coral Reef") %>%
  pull(species_richness)

nursery_richness <- comprehensive_df %>%
  filter(area_type == "Nursery") %>%
  pull(species_richness)

if (length(coral_reef_richness) > 0 && length(nursery_richness) > 0) {
  test_result <- wilcox.test(coral_reef_richness, nursery_richness)
  significance_results <- rbind(significance_results, data.frame(
    comparison = "Coral Reef vs Nursery (Species Richness)",
    n_group1 = length(coral_reef_richness),
    n_group2 = length(nursery_richness),
    test = "Mann-Whitney U (Wilcox)",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(coral_reef_richness), 2),
    mean_group2 = round(mean(nursery_richness), 2)
  ))
}

# Test 2: Milimani vs Utumbi (Species Richness)
milimani_richness <- comprehensive_df %>%
  filter(location == "Milimani") %>%
  pull(species_richness)

utumbi_richness <- comprehensive_df %>%
  filter(location == "Utumbi") %>%
  pull(species_richness)

if (length(milimani_richness) > 0 && length(utumbi_richness) > 0) {
  test_result <- wilcox.test(milimani_richness, utumbi_richness)
  significance_results <- rbind(significance_results, data.frame(
    comparison = "Milimani vs Utumbi (Species Richness)",
    n_group1 = length(milimani_richness),
    n_group2 = length(utumbi_richness),
    test = "Mann-Whitney U (Wilcox)",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(milimani_richness), 2),
    mean_group2 = round(mean(utumbi_richness), 2)
  ))
}

# Test 3: Milimani vs Utumbi (MaxN)
milimani_maxn <- comprehensive_df %>%
  filter(location == "Milimani") %>%
  pull(maxn_overall)

utumbi_maxn <- comprehensive_df %>%
  filter(location == "Utumbi") %>%
  pull(maxn_overall)

if (length(milimani_maxn) > 0 && length(utumbi_maxn) > 0) {
  test_result <- wilcox.test(milimani_maxn, utumbi_maxn)
  significance_results <- rbind(significance_results, data.frame(
    comparison = "Milimani vs Utumbi (MaxN)",
    n_group1 = length(milimani_maxn),
    n_group2 = length(utumbi_maxn),
    test = "Mann-Whitney U (Wilcox)",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(milimani_maxn), 2),
    mean_group2 = round(mean(utumbi_maxn), 2)
  ))
}

# Test 4: Kruskal-Wallis für alle Baits (Species Richness)
bait_types <- unique(comprehensive_df$bait_type)
if (length(bait_types) > 2) {
  bait_groups <- list()
  for (bait in bait_types) {
    bait_groups[[bait]] <- comprehensive_df %>%
      filter(bait_type == bait) %>%
      pull(species_richness)
  }
  
  # Kruskal-Wallis Test
  test_result <- kruskal.test(comprehensive_df$species_richness, comprehensive_df$bait_type)
  significance_results <- rbind(significance_results, data.frame(
    comparison = paste("All Baits (", length(bait_types), ") - Species Richness", sep=""),
    n_group1 = nrow(comprehensive_df),
    n_group2 = length(bait_types),
    test = "Kruskal-Wallis H",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(comprehensive_df$species_richness), 2),
    mean_group2 = ""
  ))
}

# Test 5: Paarweise Bait-Vergleiche
bait_list <- sort(unique(comprehensive_df$bait_type))
for (i in 1:min(5, length(bait_list))) {
  for (j in (i+1):min(5, length(bait_list))) {
    group1 <- comprehensive_df %>%
      filter(bait_type == bait_list[i]) %>%
      pull(species_richness)
    
    group2 <- comprehensive_df %>%
      filter(bait_type == bait_list[j]) %>%
      pull(species_richness)
    
    if (length(group1) > 0 && length(group2) > 0) {
      test_result <- wilcox.test(group1, group2)
      significance_results <- rbind(significance_results, data.frame(
        comparison = paste(bait_list[i], " vs ", bait_list[j], sep=""),
        n_group1 = length(group1),
        n_group2 = length(group2),
        test = "Mann-Whitney U (Wilcox)",
        statistic = round(test_result$statistic, 4),
        p_value = round(test_result$p.value, 6),
        significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
        mean_group1 = round(mean(group1), 2),
        mean_group2 = round(mean(group2), 2)
      ))
    }
  }
}

# Test 6: Shannon Diversity Index - Coral Reef vs Nursery
coral_reef_shannon <- comprehensive_df %>%
  filter(area_type == "Coral Reef") %>%
  pull(shannon_h)

nursery_shannon <- comprehensive_df %>%
  filter(area_type == "Nursery") %>%
  pull(shannon_h)

if (length(coral_reef_shannon) > 0 && length(nursery_shannon) > 0) {
  test_result <- wilcox.test(coral_reef_shannon, nursery_shannon)
  significance_results <- rbind(significance_results, data.frame(
    comparison = "Coral Reef vs Nursery (Shannon Diversity)",
    n_group1 = length(coral_reef_shannon),
    n_group2 = length(nursery_shannon),
    test = "Mann-Whitney U (Wilcox)",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(coral_reef_shannon), 3),
    mean_group2 = round(mean(nursery_shannon), 3)
  ))
}

print(significance_results)

# ===== 10. SPEICHERE ERGEBNISSE =====

dir.create("results", showWarnings = FALSE)

# Haupttabelle
write.csv(comprehensive_df, "results/comprehensive_analysis_table.csv", row.names = FALSE)
cat("\n\nHaupttabelle gespeichert: results/comprehensive_analysis_table.csv\n")

# Bereichs-Statistiken
write.csv(stats_by_area, "results/statistics_by_area.csv", row.names = FALSE)
cat("Bereichs-Statistiken gespeichert: results/statistics_by_area.csv\n")

# Standort-Statistiken
write.csv(stats_by_location, "results/statistics_by_location.csv", row.names = FALSE)
cat("Standort-Statistiken gespeichert: results/statistics_by_location.csv\n")

# Köder-Statistiken
write.csv(bait_stats, "results/statistics_by_bait.csv", row.names = FALSE)
cat("Köder-Statistiken gespeichert: results/statistics_by_bait.csv\n")

# Signifikanz-Tests
write.csv(significance_results, "results/statistical_significance_tests.csv", row.names = FALSE)
cat("Signifikanz-Tests gespeichert: results/statistical_significance_tests.csv\n")

cat("\n===== ANALYSE ABGESCHLOSSEN =====\n\n")
